File: client/java/test/krpc/client/TcpIpConnectionTest.java
package krpc.client;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.channels.ServerSocketChannel;

/**
 * The connection carried over TCP/IP, against a server the test listens on itself rather
 * than a kRPC server.
 */
public class TcpIpConnectionTest extends ConnectionTestCase {

  @Override
  protected ServerSocketChannel listen() throws IOException {
    // Port zero, so the system picks one nothing else is using
    return ServerSocketChannel.open().bind(new InetSocketAddress("localhost", 0));
  }

  @Override
  protected Connection connect(String name) throws IOException {
    int port = ((InetSocketAddress) getServer().getAddress()).getPort();
    return Connection.newInstance(name, "localhost", port, port);
  }
}

File: client/csharp/test/TCPIPConnectionTest.cs
using System.Net;
using System.Net.Sockets;
using NUnit.Framework;

namespace KRPC.Client.Test
{
    /// <summary>
    /// The connection carried over TCP/IP, against a server the test listens on itself
    /// rather than a kRPC server.
    /// </summary>
    [TestFixture]
    public class TCPIPConnectionTest : ConnectionTestCase
    {
        protected override Socket Listen ()
        {
            var socket = new Socket (
                AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            // Port zero, so the system picks one nothing else is using
            socket.Bind (new IPEndPoint (IPAddress.Loopback, 0));
            return socket;
        }

        protected override Connection Connect (string name)
        {
            var endPoint = (IPEndPoint)Server.EndPoint;
            return new Connection (
                name, IPAddress.Loopback, endPoint.Port, endPoint.Port);
        }
    }
}
